File: Divide an Array Into Subarrays With Minimum Cost I/Divide an Array Into Subarrays With Minimum Cost I.py
# Solution for Divide an Array Into Subarrays With Minimum Cost I in PY

class Solution:
    def minimumCost(self, nums: List[int]) -> int:

        # Cost of the first subarray (always nums[0])
        total_cost = nums[0]

        smallest = float('inf')
        second_smallest = float('inf')

        # Find two smallest values in nums[1:]
        for num in nums[1:]:
            if num < smallest:
                second_smallest = smallest
                smallest = num
            elif num < second_smallest:
                second_smallest = num

        return total_cost + smallest + second_smallest

File: Divide an Array Into Subarrays With Minimum Cost I/Divide an Array Into Subarrays With Minimum Cost I.cpp
// Solution for Divide an Array Into Subarrays With Minimum Cost I in CPP

class Solution {
public:
    int minimumCost(vector<int>& nums) {

        /*
            Observation:
            - We must divide the array into exactly 3 contiguous subarrays.
            - The "cost" of each subarray is its first element.
            - The first subarray MUST start at index 0,
              so nums[0] is always included in the answer.
        */

        int totalCost = nums[0];  // Cost of the first subarray

        /*
            We now need to choose the starting elements
            of the 2nd and 3rd subarrays.

            Since subarrays are contiguous and disjoint:
            - Their starting indices must be somewhere after index 0.
            - To minimize total cost, we just need the
              two smallest values from nums[1 ... n-1].
        */

        int smallest = INT_MAX;        // smallest value in nums[1..]
        int secondSmallest = INT_MAX;  // second smallest value in nums[1..]

        for (int i = 1; i < nums.size(); i++) {

            /*
                Update the two smallest values seen so far.
                Classic "find 2 minimum elements" pattern.
            */
            if (nums[i] < smallest) {
                secondSmallest = smallest;
                smallest = nums[i];
            } 
            else if (nums[i] < secondSmallest) {
                secondSmallest = nums[i];
            }
        }

        /*
            Final answer:
            - cost of first subarray  -> nums[0]
            - cost of second subarray -> smallest
            - cost of third subarray  -> secondSmallest
        */
        return totalCost + smallest + secondSmallest;
    }
};
